jmeter-service/testrunner.go

```diff
@@ -29,63 +36,99 @@ func NewTestRunner(eventSender *keptnv2.HTTPEventSender) *TestRunner {
+	val := ctx.Value(gracefulShutdownKey)
+	if val != nil {
+		if wg, ok := val.(*sync.WaitGroup); ok {
+			wg.Add(1)
 		}
 	}
-	msg := fmt.Sprintf("Tests for %s with status = %s. %v", testInfo.TestStrategy, strconv.FormatBool(res), testInfo)
-	if !res {
-		if err := tr.sendTestsFinishedEvent(testInfo, testStartedAt, msg, keptnv2.ResultFailed); err != nil {
-			logger.WithError(err).Errorf(errMsgSendFinishedEvent, testInfo)
+
+	resChan := make(chan TestResult, 1)
+	go tr.runTests(testInfo, jmeterConfig, resChan)
+	go tr.sendTestResult(ctx, testInfo, resChan, testStartedAt)
+
+	return nil
+}
+
+func (tr *TestRunner) sendTestResult(ctx context.Context, testInfo TestInfo, resChan chan TestResult, testStartedAt time.Time) {
+
+	defer func() {
+		val := ctx.Value(gracefulShutdownKey)
+		if val == nil {
+			return
 		}
-	} else {
-		if err := tr.sendTestsFinishedEvent(testInfo, testStartedAt, msg, keptnv2.ResultPass); err != nil {
-			logger.WithError(err).Errorf(errMsgSendFinishedEvent, testInfo)
+		if wg, ok := val.(*sync.WaitGroup); ok {
+			wg.Done()
+		}
+	}()
+
+	select {
+	case result := <-resChan:
+		logger.Info("Sending result ", testInfo, result.res)
+		if result.err != nil {
+			if err := tr.sendErroredTestsFinishedEvent(testInfo, testStartedAt, result.err.Error()); err != nil {
+				logger.WithError(err).Errorf(errMsgSendFinishedEvent, testInfo)
+			}
+		}
+		msg := fmt.Sprintf("Tests for %s with status = %s. %v", testInfo.TestStrategy, strconv.FormatBool(result.res), testInfo)
+		if !result.res {
+			if err := tr.sendTestsFinishedEvent(testInfo, testStartedAt, msg, keptnv2.ResultFailed); err != nil {
+				logger.WithError(err).Errorf(errMsgSendFinishedEvent, testInfo)
+			}
+		} else {
+			if err := tr.sendTestsFinishedEvent(testInfo, testStartedAt, msg, keptnv2.ResultPass); err != nil {
+				logger.WithError(err).Errorf(errMsgSendFinishedEvent, testInfo)
+			}
 		}
+	case <-ctx.Value(keptnQuit).(chan os.Signal): /// this avoids to answer to context.Done from cloud event lib
+		logger.Errorf("Terminated, sending test finished event %v", ctx.Err())
+		if err := tr.sendErroredTestsFinishedEvent(testInfo, testStartedAt, "received a SIGTERM/SIGINT, jmeter terminated before the end of the test"); err != nil {
+			logger.Error(fmt.Sprintf("Error sending test finished event: %s", err.Error()) + ". " + testInfo.String())
+		}
+
 	}
-	return nil
 }
 
-func (tr *TestRunner) runTests(testInfo TestInfo, jmeterConf *JMeterConf) (bool, error) {
+func (tr *TestRunner) runTests(testInfo TestInfo, jmeterConf *JMeterConf, resChan chan TestResult) {
 	var testStrategy = strings.ToLower(testInfo.TestStrategy)
+	res := false
 
 	if testStrategy == "" {
 		logger.Info("No testStrategy specified therefore skipping further test execution and sending back success")
-		return true, nil
 	}
 
 	testStrategyWorkload, err := getWorkloadForStrategy(jmeterConf, testStrategy)
 	if err != nil {
-		return false, err
+		logger.Error(err.Error())
```

Comment:
please add a message in front of the error. Also, was the removal of the `return` at this point intended (see also the following lines where previously an error was returned)?

Reply:
yes since I propagate the error in the res channel line 131